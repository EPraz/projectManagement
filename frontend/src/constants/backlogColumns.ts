// Actualiza tu interfaz para incluir una propiedad de orden
export interface BacklogColumnsProps {
  key: string;
  label: string;
  order: number; // Añadimos una propiedad de orden explícita
  width?: string; // Opcional: ancho personalizado
}

// Actualiza tus columnas con la propiedad de orden
export const backlogColumns: BacklogColumnsProps[] = [
  { key: "order", label: "Order", order: 1, width: "10%" },
  { key: "id", label: "ID", order: 2, width: "10%" },
  { key: "state", label: "State", order: 3, width: "15%" },
  { key: "title", label: "Title", order: 4 },
  { key: "type", label: "Type", order: 5 },
  { key: "priority", label: "Priority", order: 6, width: "10%" },
  { key: "assignedTo", label: "Assigned To", order: 7 },
  { key: "sprint", label: "Sprint", order: 8 },
  { key: "dueDate", label: "Due Date", order: 9 },
  { key: "estimatedHours", label: "Est. Hours", order: 10, width: "8%" },
  { key: "remainingHours", label: "Rem. Hours", order: 11, width: "8%" },
  { key: "isBlocked", label: "Blocking", order: 12 },
  { key: "tags", label: "Tags", order: 13 },
  { key: "actions", label: "Actions", order: 14, width: "5%" },
];

// Columnas inicialmente visibles
export const initialBacklogColumns: string[] = [
  "order",
  "id",
  "state",
  "title",
  "priority",
  "tags",
  "actions",
];
