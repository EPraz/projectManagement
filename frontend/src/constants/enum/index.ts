export * from "./Role";
export * from "./TicketPriority";
export * from "./TicketType";
