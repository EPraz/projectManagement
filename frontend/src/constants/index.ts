export * from "./taskStatus";
export * from "./ticketStatus";
export * from "./enum";
export * from "./menuItems";
export * from "./tabPanels";
export * from "./backlogColumns";
export * from "./apiConstant";
