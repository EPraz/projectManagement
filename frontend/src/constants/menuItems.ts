import DashboardIcon from "@mui/icons-material/Dashboard";
import ViewKanbanIcon from "@mui/icons-material/ViewKanban";
import AssessmentIcon from "@mui/icons-material/Assessment";
import CategoryIcon from "@mui/icons-material/Category";
import PeopleIcon from "@mui/icons-material/People";
import SettingsIcon from "@mui/icons-material/Settings";

export interface MenuItem {
  text: string;
  icon: React.ElementType;
  path: string;
  disabled?: boolean;
}

export const MENU_ITEMS: MenuItem[] = [
  {
    text: "Overview",
    icon: DashboardIcon,
    path: "overview",
  },
  {
    text: "Board",
    icon: ViewKanbanIcon,
    path: "board",
  },
  {
    text: "Retrospective",
    icon: AssessmentIcon,
    path: "retrospective",
  },
  {
    text: "Epics & Features",
    icon: CategoryIcon,
    path: "epics",
    disabled: true,
  },

  {
    text: "Team Members",
    icon: PeopleIcon,
    path: "teammembers",
  },
  {
    text: "Settings",
    icon: SettingsIcon,
    path: "settings",
  },
];

export interface MainTab {
  text: string;
  path: string;
}

export const MAIN_TABS: MainTab[] = [
  {
    text: "Overview",
    path: "overview",
  },
  {
    text: "Board",
    path: "board",
  },
  {
    text: "Members",
    path: "members",
  },
  {
    text: "Files",
    path: "files",
  },
  {
    text: "Timeline",
    path: "timeline",
  },
  {
    text: "Reports",
    path: "reports",
  },
];

export interface Project {
  name: string;
  count: number;
  color: string;
}

export const PROJECTS: Project[] = [
  { name: "ShieldSafe", count: 5, color: "#1E8E3E" },
  { name: "Quantum", count: 2, color: "#1A73E8" },
  { name: "AeroSquare", count: 7, color: "#F9AB00" },
];

export const cardVariants = {
  hidden: { opacity: 0, y: 20 },
  visible: { opacity: 1, y: 0 },
};
