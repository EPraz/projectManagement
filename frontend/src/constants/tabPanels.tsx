import ViewKanbanIcon from "@mui/icons-material/ViewKanban";
import FormatListBulletedIcon from "@mui/icons-material/FormatListBulleted";
import GroupIcon from "@mui/icons-material/Group";
import FlagIcon from "@mui/icons-material/Flag";

import { Taskboard } from "../pages";
import BacklogBoard from "../pages/board/backlogBoard/BacklogBoard";
import CapacityBoard from "../pages/board/capacityBoard/CapacityBoard";
import GoalBoard from "../pages/board/goalBoard/GoalBoard";

interface TabPanel {
  label: string;
  icon: JSX.Element;
  component: () => JSX.Element;
}

export const TabPanels: TabPanel[] = [
  { label: "Board", icon: <ViewKanbanIcon />, component: () => <Taskboard /> },
  {
    label: "Backlog",
    icon: <FormatListBulletedIcon />,
    component: () => <BacklogBoard />,
  },
  {
    label: "Capacity",
    icon: <GroupIcon />,
    component: () => <CapacityBoard />,
  },
  { label: "Goals", icon: <FlagIcon />, component: () => <GoalBoard /> },
];
