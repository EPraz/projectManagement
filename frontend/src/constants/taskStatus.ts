export const TASK_STATUSES = [
  // to be deleted
  "To Do",
  "In Progress",
  "Blocked",
  "Review",
  "Testing",
  "Done",
];
