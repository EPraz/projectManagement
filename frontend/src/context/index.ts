export * from "./sprintContext";
export * from "./apiContext";
export * from "./projectContext";
export * from './snackbarContext'