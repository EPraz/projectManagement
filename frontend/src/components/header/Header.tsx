import React, { useMemo } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import { MENU_ITEMS } from "../../constants/menuItems";
import {
  HeaderContainer,
  ProjectTitle,
  StyledTab,
  StyledTabs,
  TabsBox,
  TopBar,
} from "./Header.styles";
import { HeaderProps } from "../../types";
import UserMenu from "./userMenu/UserMenu";
import { Breadcrumbs, Link } from "@mui/material";
import HomeIcon from "@mui/icons-material/Home";

const Header: React.FC<HeaderProps> = ({ projectName, projectId }) => {
  const navigate = useNavigate();
  const location = useLocation();
  const pathSegments = location.pathname.split("/");
  const currentTab = pathSegments[pathSegments.length - 1];

  const handleTabChange = (event: React.SyntheticEvent, newValue: string) => {
    event.stopPropagation();
    if (currentTab !== newValue) {
      navigate(`/projects/${projectId}/${newValue}`, { replace: true });
    }
  };

  const tabs = useMemo(
    () =>
      MENU_ITEMS.map((tab) => (
        <StyledTab
          key={tab.path}
          label={
            <TabsBox>
              {tab.icon && <tab.icon sx={{ fontSize: 18 }} />}
              {tab.text}
            </TabsBox>
          }
          value={tab.path}
          disableRipple
          disabled={tab?.disabled}
        />
      )),
    []
  );

  const handleBreadcrumbClick = () => {
    // Navegar a la URL base del proyecto
    navigate(`/projects`, { replace: true });
  };

  return (
    <HeaderContainer>
      <TopBar>
        <Breadcrumbs aria-label="breadcrumb">
          <Link
            color="inherit"
            onClick={handleBreadcrumbClick}
            underline="hover"
            sx={{ cursor: "pointer" }}
          >
            <HomeIcon sx={{ mr: 0.5 }} fontSize="inherit" />
          </Link>
          <ProjectTitle>{projectName}</ProjectTitle>
        </Breadcrumbs>
      </TopBar>
      <StyledTabs
        value={currentTab}
        onChange={handleTabChange}
        variant="scrollable"
        scrollButtons="auto"
      >
        {tabs}
      </StyledTabs>
      <UserMenu />
    </HeaderContainer>
  );
};

export default Header;
