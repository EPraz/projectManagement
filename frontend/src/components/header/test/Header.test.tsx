// Header.test.tsx
import { render, screen, fireEvent, act } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Header from "../Header";
import { MENU_ITEMS } from "../../../constants";
import { AuthProvider } from "../../../context";

// Mockear useNavigate y useLocation
const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => {
  const actual = jest.requireActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
    useLocation: () => ({
      pathname: "/projects/123/overview",
    }),
  };
});

jest.mock("../../../constants/apiConstant", () => ({
  API_URL: "http://localhost:3000",
}));

describe("Header Component", () => {
  const projectName = "Test Project";
  const projectId = "123";

  beforeEach(() => {
    mockNavigate.mockClear();
  });

  afterAll(() => {
    jest.clearAllTimers();
  });

  const renderHeader = () =>
    render(
      <AuthProvider>
        <MemoryRouter>
          <Header projectName={projectName} projectId={projectId} />
        </MemoryRouter>
      </AuthProvider>
    );

  it("match snapshot", async () => {
    await act(async () => {
      const { asFragment } = renderHeader();
      expect(asFragment).toMatchSnapshot();
    });
  });

  it("renders project title", async () => {
    await act(async () => {
      renderHeader();
    });
    expect(screen.getByText(projectName)).toBeInTheDocument();
  });

  it("renders tabs based on MENU_ITEMS", async () => {
    await act(async () => {
      renderHeader();
    });
    MENU_ITEMS.forEach((item) => {
      expect(screen.getByText(item.text)).toBeInTheDocument();
    });
  });

  it("navigates to a new tab when a different tab is clicked", async () => {
    await act(async () => {
      renderHeader();
    });

    // currentTab se obtiene de la URL: en nuestro mock es "overview"
    // Escogemos una tab cuyo valor (path) sea diferente, por ejemplo, "dashboard"
    const tabToClick = screen.getByText(/board/i);
    fireEvent.click(tabToClick);

    expect(mockNavigate).toHaveBeenCalledWith(`/projects/${projectId}/board`, {
      replace: true,
    });
  });

  it("does not navigate if the selected tab is the current tab", async () => {
    await act(async () => {
      renderHeader();
    });
    // En nuestro mock useLocation, la tab actual es "overview"
    const tab = screen.getByText(/overview/i);
    fireEvent.click(tab);
    expect(mockNavigate).not.toHaveBeenCalled();
  });
});
