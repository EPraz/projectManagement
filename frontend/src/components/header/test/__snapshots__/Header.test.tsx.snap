// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component match snapshot 1`] = `[Function]`;
